File: experiments/mathexpr-spec-python-sonnet/mathexpr.py
"""Math expression parser and evaluator."""

from dataclasses import dataclass
from typing import Union


# ============================================================================
# Types
# ============================================================================

@dataclass
class Token:
    """A lexical token with a kind and string value."""
    kind: str
    value: str


@dataclass
class NumberLiteral:
    """AST node: a numeric literal."""
    type: str = "number"
    value: float = 0.0

    def __init__(self, value: float):
        self.type = "number"
        self.value = value


@dataclass
class UnaryExpr:
    """AST node: unary operator applied to an operand."""
    type: str = "unary"
    op: str = ""
    operand: 'AstNode' = None

    def __init__(self, op: str, operand: 'AstNode'):
        self.type = "unary"
        self.op = op
        self.operand = operand


@dataclass
class BinaryExpr:
    """AST node: binary operator applied to left and right operands."""
    type: str = "binary"
    op: str = ""
    left: 'AstNode' = None
    right: 'AstNode' = None

    def __init__(self, op: str, left: 'AstNode', right: 'AstNode'):
        self.type = "binary"
        self.op = op
        self.left = left
        self.right = right


AstNode = Union[NumberLiteral, UnaryExpr, BinaryExpr]


# ============================================================================
# Tokenizer
# ============================================================================

def tokenize(input: str) -> list[Token]:
    """
    Convert a math expression string into a sequence of tokens.
    
    Skips whitespace, parses numbers (with optional decimal point),
    operators (+, -, *, /, %, **), and parentheses.
    
    Raises ValueError on unrecognized characters.
    """
    tokens = []
    i = 0
    length = len(input)
    
    while i < length:
        ch = input[i]
        
        # Skip whitespace
        if ch in ' \t\n\r':
            i += 1
            continue
        
        # Number: digits and at most one decimal point
        if ch.isdigit() or ch == '.':
            start = i
            has_dot = ch == '.'
            i += 1
            
            while i < length:
                ch = input[i]
                if ch.isdigit():
                    i += 1
                elif ch == '.':
                    if has_dot:
                        raise ValueError(f"Unexpected character {ch}")
                    has_dot = True
                    i += 1
                else:
                    break
            
            tokens.append(Token("number", input[start:i]))
            continue
        
        # Two-character operator: **
        if ch == '*' and i + 1 < length and input[i + 1] == '*':
            tokens.append(Token("power", "**"))
            i += 2
            continue
        
        # Single-character operators and parens
        if ch == '+':
            tokens.append(Token("plus", "+"))
            i += 1
        elif ch == '-':
            tokens.append(Token("minus", "-"))
            i += 1
        elif ch == '*':
            tokens.append(Token("star", "*"))
            i += 1
        elif ch == '/':
            tokens.append(Token("slash", "/"))
            i += 1
        elif ch == '%':
            tokens.append(Token("percent", "%"))
            i += 1
        elif ch == '(':
            tokens.append(Token("lparen", "("))
            i += 1
        elif ch == ')':
            tokens.append(Token("rparen", ")"))
            i += 1
        else:
            raise ValueError(f"Unexpected character {ch} at position {i}")
    
    return tokens


# ============================================================================
# Parser
# ============================================================================

class Parser:
    """Recursive descent parser with precedence climbing."""
    
    def __init__(self, tokens: list[Token]):
        self.tokens = tokens
        self.pos = 0
    
    def current(self) -> Token | None:
        """Get current token without consuming it."""
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return None
    
    def consume(self) -> Token:
        """Consume and return current token."""
        if self.pos >= len(self.tokens):
            raise ValueError("Unexpected end of input")
        token = self.tokens[self.pos]
        self.pos += 1
        return token
    
    def expect(self, kind: str) -> Token:
        """Consume token of expected kind or raise error."""
        if self.pos >= len(self.tokens):
            raise ValueError(f"Expected {kind}")
        token = self.tokens[self.pos]
        if token.kind != kind:
            raise ValueError(f"Expected {kind}")
        self.pos += 1
        return token
    
    def parse_expression(self) -> AstNode:
        """Parse entry point: additive expression (lowest precedence)."""
        if not self.tokens:
            raise ValueError("Unexpected end of input")
        
        expr = self.parse_additive()
        
        # Check for trailing tokens
        if self.pos < len(self.tokens):
            raise ValueError("Unexpected token after expression")
        
        return expr
    
    def parse_additive(self) -> AstNode:
        """Parse + and - (left-associative, precedence 1)."""
        left = self.parse_multiplicative()
        
        while True:
            token = self.current()
            if token and token.kind in ("plus", "minus"):
                op = "+" if token.kind == "plus" else "-"
                self.consume()
                right = self.parse_multiplicative()
                left = BinaryExpr(op, left, right)
            else:
                break
        
        return left
    
    def parse_multiplicative(self) -> AstNode:
        """Parse *, /, % (left-associative, precedence 2)."""
        left = self.parse_power()
        
        while True:
            token = self.current()
            if token and token.kind in ("star", "slash", "percent"):
                op = {"star": "*", "slash": "/", "percent": "%"}[token.kind]
                self.consume()
                right = self.parse_power()
                left = BinaryExpr(op, left, right)
            else:
                break
        
        return left
    
    def parse_power(self) -> AstNode:
        """Parse ** (right-associative, precedence 3)."""
        left = self.parse_unary()
        
        token = self.current()
        if token and token.kind == "power":
            self.consume()
            # Right-associative: recurse into parse_power
            right = self.parse_power()
            return BinaryExpr("**", left, right)
        
        return left
    
    def parse_unary(self) -> AstNode:
        """Parse unary - (prefix, precedence 4)."""
        token = self.current()
        
        if token and token.kind == "minus":
            self.consume()
            operand = self.parse_unary()  # Can chain: --5
            return UnaryExpr("-", operand)
        
        return self.parse_atom()
    
    def parse_atom(self) -> AstNode:
        """Parse atoms: numbers and parenthesized expressions (precedence 5)."""
        token = self.current()
        
        if not token:
            raise ValueError("Unexpected end of input")
        
        # Number literal
        if token.kind == "number":
            self.consume()
            value = float(token.value)
            return NumberLiteral(value)
        
        # Parenthesized expression
        if token.kind == "lparen":
            self.consume()
            expr = self.parse_additive()  # Recurse to lowest precedence
            self.expect("rparen")
            return expr
        
        raise ValueError(f"Unexpected token: {token.kind}")


def parse(tokens: list[Token]) -> AstNode:
    """Parse a token sequence into an AST using recursive descent."""
    parser = Parser(tokens)
    return parser.parse_expression()


# ============================================================================
# Evaluator
# ============================================================================

def evaluate(ast: AstNode) -> float:
    """
    Evaluate an AST node by recursive tree walk.
    
    Raises ValueError on division by zero or modulo by zero.
    """
    if isinstance(ast, NumberLiteral):
        return ast.value
    
    if isinstance(ast, UnaryExpr):
        operand_val = evaluate(ast.operand)
        if ast.op == "-":
            return -operand_val
        raise ValueError(f"Unknown unary operator: {ast.op}")
    
    if isinstance(ast, BinaryExpr):
        left_val = evaluate(ast.left)
        right_val = evaluate(ast.right)
        
        if ast.op == "+":
            return left_val + right_val
        elif ast.op == "-":
            return left_val - right_val
        elif ast.op == "*":
            return left_val * right_val
        elif ast.op == "/":
            if right_val == 0:
                raise ValueError("Division by zero")
            return left_val / right_val
        elif ast.op == "%":
            if right_val == 0:
                raise ValueError("Modulo by zero")
            return left_val % right_val
        elif ast.op == "**":
            return left_val ** right_val
        else:
            raise ValueError(f"Unknown binary operator: {ast.op}")
    
    raise ValueError(f"Unknown AST node type: {type(ast)}")


# ============================================================================
# End-to-End
# ============================================================================

def calc(expression: str) -> float:
    """
    End-to-end: parse and evaluate a math expression string.
    
    Raises ValueError on empty/whitespace-only input or evaluation errors.
    """
    # Check for empty/whitespace-only input
    if not expression or expression.strip() == "":
        raise ValueError("Empty expression")
    
    tokens = tokenize(expression)
    ast = parse(tokens)
    return evaluate(ast)

File: research/experiments/mathexpr-spec-go-haiku/mathexpr.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
	"unicode"
)

// Token represents a lexical token
type Token struct {
	Kind  string
	Value string
}

// Token kinds
const (
	TokenNumber  = "number"
	TokenPlus    = "plus"
	TokenMinus   = "minus"
	TokenStar    = "star"
	TokenSlash   = "slash"
	TokenPercent = "percent"
	TokenPower   = "power"
	TokenLParen  = "lparen"
	TokenRParen  = "rparen"
)

// AstNode represents a node in the abstract syntax tree
type AstNode interface {
	// marker interface
}

// NumberLiteral represents a numeric constant
type NumberLiteral struct {
	Value float64
}

// UnaryExpr represents a unary operation (only unary minus is supported)
type UnaryExpr struct {
	Op      string  // "-"
	Operand AstNode
}

// BinaryExpr represents a binary operation
type BinaryExpr struct {
	Op    string  // "+", "-", "*", "/", "%", "**"
	Left  AstNode
	Right AstNode
}

// Tokenize converts a string into a sequence of tokens
func Tokenize(input string) ([]Token, error) {
	var tokens []Token
	i := 0

	for i < len(input) {
		ch := input[i]

		// Skip whitespace
		if ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' {
			i++
			continue
		}

		// Parse number
		if unicode.IsDigit(rune(ch)) || (ch == '.' && i+1 < len(input) && unicode.IsDigit(rune(input[i+1]))) || (ch == '.' && i+1 == len(input)) {
			start := i
			hasDot := false

			for i < len(input) && (unicode.IsDigit(rune(input[i])) || input[i] == '.') {
				if input[i] == '.' {
					if hasDot {
						return nil, fmt.Errorf("unexpected character '.' at position %d", i)
					}
					hasDot = true
				}
				i++
			}

			tokens = append(tokens, Token{Kind: TokenNumber, Value: input[start:i]})
			continue
		}

		// Parse operators and parentheses
		switch ch {
		case '+':
			tokens = append(tokens, Token{Kind: TokenPlus, Value: "+"})
			i++
		case '-':
			tokens = append(tokens, Token{Kind: TokenMinus, Value: "-"})
			i++
		case '*':
			if i+1 < len(input) && input[i+1] == '*' {
				tokens = append(tokens, Token{Kind: TokenPower, Value: "**"})
				i += 2
			} else {
				tokens = append(tokens, Token{Kind: TokenStar, Value: "*"})
				i++
			}
		case '/':
			tokens = append(tokens, Token{Kind: TokenSlash, Value: "/"})
			i++
		case '%':
			tokens = append(tokens, Token{Kind: TokenPercent, Value: "%"})
			i++
		case '(':
			tokens = append(tokens, Token{Kind: TokenLParen, Value: "("})
			i++
		case ')':
			tokens = append(tokens, Token{Kind: TokenRParen, Value: ")"})
			i++
		default:
			return nil, fmt.Errorf("unexpected character '%c' at position %d", ch, i)
		}
	}

	return tokens, nil
}

// Parser holds state for parsing
type Parser struct {
	tokens []Token
	pos    int
}

// Parse converts a token sequence into an AST
func Parse(tokens []Token) (AstNode, error) {
	if len(tokens) == 0 {
		return nil, fmt.Errorf("unexpected end of input")
	}

	p := &Parser{tokens: tokens, pos: 0}
	expr, err := p.parseExpression()
	if err != nil {
		return nil, err
	}

	if p.pos < len(p.tokens) {
		return nil, fmt.Errorf("unexpected token after expression")
	}

	return expr, nil
}

func (p *Parser) parseExpression() (AstNode, error) {
	return p.parseAdditive()
}

func (p *Parser) parseAdditive() (AstNode, error) {
	left, err := p.parseMultiplicative()
	if err != nil {
		return nil, err
	}

	for p.pos < len(p.tokens) && (p.tokens[p.pos].Kind == TokenPlus || p.tokens[p.pos].Kind == TokenMinus) {
		op := p.tokens[p.pos].Value
		p.pos++

		right, err := p.parseMultiplicative()
		if err != nil {
			return nil, err
		}

		left = &BinaryExpr{Op: op, Left: left, Right: right}
	}

	return left, nil
}

func (p *Parser) parseMultiplicative() (AstNode, error) {
	left, err := p.parsePower()
	if err != nil {
		return nil, err
	}

	for p.pos < len(p.tokens) && (p.tokens[p.pos].Kind == TokenStar || p.tokens[p.pos].Kind == TokenSlash || p.tokens[p.pos].Kind == TokenPercent) {
		op := p.tokens[p.pos].Value
		p.pos++

		right, err := p.parsePower()
		if err != nil {
			return nil, err
		}

		left = &BinaryExpr{Op: op, Left: left, Right: right}
	}

	return left, nil
}

func (p *Parser) parsePower() (AstNode, error) {
	left, err := p.parseUnary()
	if err != nil {
		return nil, err
	}

	if p.pos < len(p.tokens) && p.tokens[p.pos].Kind == TokenPower {
		op := p.tokens[p.pos].Value
		p.pos++

		right, err := p.parsePower() // Right associative
		if err != nil {
			return nil, err
		}

		left = &BinaryExpr{Op: op, Left: left, Right: right}
	}

	return left, nil
}

func (p *Parser) parseUnary() (AstNode, error) {
	if p.pos < len(p.tokens) && p.tokens[p.pos].Kind == TokenMinus {
		p.pos++

		operand, err := p.parseUnary() // Right associative
		if err != nil {
			return nil, err
		}

		return &UnaryExpr{Op: "-", Operand: operand}, nil
	}

	return p.parseAtom()
}

func (p *Parser) parseAtom() (AstNode, error) {
	if p.pos >= len(p.tokens) {
		return nil, fmt.Errorf("unexpected end of input")
	}

	token := p.tokens[p.pos]

	if token.Kind == TokenNumber {
		p.pos++
		val, err := strconv.ParseFloat(token.Value, 64)
		if err != nil {
			return nil, fmt.Errorf("invalid number: %s", token.Value)
		}
		return &NumberLiteral{Value: val}, nil
	}

	if token.Kind == TokenLParen {
		p.pos++
		expr, err := p.parseExpression()
		if err != nil {
			return nil, err
		}

		if p.pos >= len(p.tokens) || p.tokens[p.pos].Kind != TokenRParen {
			return nil, fmt.Errorf("expected rparen")
		}

		p.pos++
		return expr, nil
	}

	return nil, fmt.Errorf("unexpected token: %s", token.Kind)
}

// Evaluate evaluates an AST node to a numeric value
func Evaluate(node AstNode) (float64, error) {
	switch n := node.(type) {
	case *NumberLiteral:
		return n.Value, nil
	case *UnaryExpr:
		operand, err := Evaluate(n.Operand)
		if err != nil {
			return 0, err
		}
		return -operand, nil
	case *BinaryExpr:
		left, err := Evaluate(n.Left)
		if err != nil {
			return 0, err
		}
		right, err := Evaluate(n.Right)
		if err != nil {
			return 0, err
		}

		switch n.Op {
		case "+":
			return left + right, nil
		case "-":
			return left - right, nil
		case "*":
			return left * right, nil
		case "/":
			if right == 0 {
				return 0, fmt.Errorf("division by zero")
			}
			return left / right, nil
		case "%":
			if right == 0 {
				return 0, fmt.Errorf("modulo by zero")
			}
			return math.Mod(left, right), nil
		case "**":
			return math.Pow(left, right), nil
		default:
			return 0, fmt.Errorf("unknown operator: %s", n.Op)
		}
	default:
		return 0, fmt.Errorf("unknown AST node type")
	}
}

// Calc is the end-to-end function: expression string -> number
func Calc(expression string) (float64, error) {
	trimmed := strings.TrimSpace(expression)
	if trimmed == "" {
		return 0, fmt.Errorf("empty expression")
	}

	tokens, err := Tokenize(expression)
	if err != nil {
		return 0, err
	}

	if len(tokens) == 0 {
		return 0, fmt.Errorf("empty expression")
	}

	ast, err := Parse(tokens)
	if err != nil {
		return 0, err
	}

	result, err := Evaluate(ast)
	if err != nil {
		return 0, err
	}

	return result, nil
}
